txscript/taproot.go

```diff
@@ -0,0 +1,642 @@
+// Copyright (c) 2013-2022 The btcsuite developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package txscript
+
+import (
+	"bytes"
+	"fmt"
+
+	secp "github.com/btcsuite/btcd/btcec/v2"
+	"github.com/btcsuite/btcd/btcec/v2/schnorr"
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"github.com/btcsuite/btcd/wire"
+)
+
+// LeafVersionMask is the mask applied to the first byte of the control block
+// to extract the committed leaf versions.
+const LeafVersionMask = 0xfe
+
+// TapscriptLeafVersion represents the various possible versions of a tapscript
+// leaf version. Leaf versions are used to define, or introduce new script
+// semantics, under the base taproot execution model.
+//
+// TODO(roasbeef): add validation here as well re proper prefix, etc?
+type TapscriptLeafVersion uint8
+
+const (
+	// BaseLeafVersion is the base tapscript leaf version. The semantics of
+	// this version are defined in BIP 342.
+	BaseLeafVersion TapscriptLeafVersion = 0xc0
+)
+
+const (
+	// ControlBlockBaseSize is the base size of a control block. This
+	// includes the initial byte for the leaf version, and then serialized
+	// schnorr public key.
+	ControlBlockBaseSize = 33
+
+	// ControlBlockNodeSize is the size of a given merkle branch hash in
+	// the control block.
+	ControlBlockNodeSize = 32
+
+	// ControlBlockMaxNodeCount is the max number of nodes that can be
+	// included in a control block. This value represents a merkle tree of
+	// depth 2^128.
+	ControlBlockMaxNodeCount = 128
+
+	// ControlBlockMaxSize is the max possible size of a control block.
+	// This simulates revealing a leaf from the largest possible tapscript
+	// tree.
+	ControlBlockMaxSize = ControlBlockBaseSize + (ControlBlockNodeSize *
+		ControlBlockMaxNodeCount)
+)
+
+var (
+	// TaprootTagSigHash is the message tag prefix used to compute the
+	// sighash used to validate taproot schnorr signatures.
+	TaprootTagSigHash = []byte("TapSighhash")
```

Comment:
@Roasbeef Consider enforcing ASCII encoding to match TaggedHash in Bitcoin's hash.cpp. Bitcoin's hash.cpp assumes one byte per character. UTF-8 encoding does lead to different hashes and to invalid signatures. An alternative is 
`	TaprootTagSigHash = []byte{0x54, 0x61, 0x70, 0x53, 0x69, 0x67, 0x68, 0x61, 0x73, 0x68} // "TapSighhash"
`